server/src/main/java/org/elasticsearch/common/breaker/CircuitBreaker.java

```diff
@@ -107,6 +107,14 @@ public static Type parseValue(String value) {
     String getName();
 
     default long getFree() {
-        return getLimit() - getUsed();
+        var limit = getLimit();
+        if (limit == -1) {
+            // -1 disables breaking
+            return Long.MAX_VALUE;
+        } else if (limit == 0) {
+            // 0 disables any data usage
+            return 0;
+        }
```

Comment:
Could we replace special value in the `addSettingsUpdateConsumer` in order to make it once and not on each `getFree` call?

Reply:
Good point, pushed a fixup.